File: webapp/lib/hero_ability.rb
# frozen_string_literal: true

require_relative "room"

# Hero special abilities: damage modifiers applied to each encounter during the
# crawl. Each ability responds to `reduced(encounter, damage)` (the damage taken,
# never below zero) and has a `scope`:
#   :party — protects whoever takes the hit, as long as this hero is alive in the
#            party (Cleric, Mage, Rogue).
#   :self  — only reduces damage this hero personally takes (Barbarian).
#
# Looked up by hero id via HeroAbility.lookup; heroes without an entry take full
# damage (the null ability). Use HeroAbility.damage_taken to combine a party's
# abilities for a single hit.
module HeroAbility
  # No reduction.
  class Null
    def scope = :self

    def reduced(_encounter, damage)
      damage
    end
  end

  # Halve all damage, rounded up — self only (Barbarian).
  class HalveRoundedUp
    def scope = :self

    def reduced(_encounter, damage)
      (damage / 2.0).ceil
    end
  end

  # Reduce damage from any encounter (room OR boss) carrying a given bait by a
  # fixed amount, for the whole party (Cleric: undead −4, Mage: power −4).
  class ReduceBaitDamage
    def initialize(bait, amount)
      @bait = bait
      @amount = amount
    end

    def scope = :party

    def reduced(encounter, damage)
      return damage unless encounter.bait.count(@bait).positive?

      [damage - @amount, 0].max
    end
  end

  # Reduce damage from trap-type rooms by a fixed amount, for the whole party
  # (Rogue: −2).
  class ReduceTrapDamage
    def initialize(amount)
      @amount = amount
    end

    def scope = :party

    def reduced(encounter, damage)
      return damage unless encounter.respond_to?(:type) && encounter.type.downcase.include?("trap")

      [damage - @amount, 0].max
    end
  end

  NULL = Null.new

  REGISTRY = {
    "hero_barbarian" => HalveRoundedUp.new,
    "hero_cleric"    => ReduceBaitDamage.new(Bait::UNDEAD, 4),
    "hero_mage"      => ReduceBaitDamage.new(Bait::POWER, 4),
    "hero_rogue"     => ReduceTrapDamage.new(2)
  }.freeze

  # The ability for a hero, or the null ability if it has none.
  def self.lookup(hero)
    REGISTRY.fetch(hero.id, NULL)
  end

  # Damage a target hero actually takes from an encounter, given the heroes still
  # alive in its party. Every alive member's :party aura applies; the target's
  # own :self ability applies last. `base` defaults to the encounter's printed
  # damage but can be overridden (e.g. a boss's points bonus).
  def self.damage_taken(target, encounter, alive_members, base = encounter.damage)
    damage = base
    alive_members.each do |member|
      ability = lookup(member)
      damage = ability.reduced(encounter, damage) if ability.scope == :party
    end
    own = lookup(target)
    damage = own.reduced(encounter, damage) if own.scope == :self
    damage
  end
end

File: android/app/src/main/java/com/dungeonboss/game/HeroAbility.kt
package com.dungeonboss.game

import com.dungeonboss.model.Encounter
import com.dungeonboss.model.Hero
import kotlin.math.ceil

/**
 * Hero damage modifiers, read from each hero's **data fields** (no per-id code):
 *
 *   party aura — `partyDamageReduction` (levelled), a flat reduction applied to
 *                whichever member is hit, while the granting hero is alive in the
 *                party, gated by the bait/room-type filters. Auras stack across
 *                members. (Cleric, Mage, Rogue.)
 *   self mult  — `selfDamageMultiplier`, applied (rounded up) only to the damage
 *                the hero personally takes, unconditionally. (Barbarian: 0.5.)
 *
 * [damageTaken] applies the party auras first, then the target's self multiplier
 * last; never below zero. Mirrors docs/pseudocode.md (HeroAbility).
 */
object HeroAbility {

    /** Does [member]'s party aura apply to this encounter (both filters must hold)? */
    private fun filterMatches(member: Hero, encounter: Encounter): Boolean {
        val baitFilter = member.damageBaitFilter
        if (baitFilter != null && encounter.bait.count(baitFilter) == 0) return false
        val typeFilter = member.damageRoomTypeFilter?.lowercase()
        if (typeFilter != null) {
            val matchesType = when (typeFilter) {
                "trap" -> encounter.trap()
                "creature", "monster" -> encounter.creature()
                else -> encounter.type?.lowercase()?.contains(typeFilter) ?: false
            }
            if (!matchesType) return false
        }
        return true
    }

    /** The levelled flat reduction [member] grants against this encounter (0 if filtered out). */
    private fun partyReduction(member: Hero, encounter: Encounter): Int =
        if (filterMatches(member, encounter)) member.partyReduction else 0

    /**
     * Damage a target hero actually takes from an encounter, given the heroes
     * still alive in its party. Every alive member's party aura applies first;
     * the target's own self multiplier (rounded up) applies last. Never below zero.
     */
    fun damageTaken(target: Hero, encounter: Encounter, aliveMembers: List<Hero>, base: Int = encounter.damage): Int {
        var damage = base
        for (member in aliveMembers) {
            damage = maxOf(damage - partyReduction(member, encounter), 0)
        }
        damage = ceil(damage * target.selfDamageMultiplier).toInt()
        return maxOf(damage, 0)
    }
}
